### ISSUE TITLE
Some Text Not Support in iOS, such as chinese "中华人民共和国"  "missing glyphs" 

### ISSUE THREAD

**Issue by AceDong0803** [tangrams/tangram-es#2188]
Some Text rendering Not Support in iOS, such as chinese "中华人民共和国" the code show missing glyphs。can't render the label.

**Comment by AceDong0803** [comment]
the word "人" or "国" missing glyphs in iOS

**Comment by AceDong0803** [comment]
I try only support  Traditional Chinese Characters ，Not support Simplified Chinese Characters，so how can I support  Simplified Chinese Characters

**Comment by matteblair** [comment]
Hi, please verify some info so that I can help figure this out.

1. What version of Tangram ES are you using?
2. What version of iOS are you using?
3. What scene file are you using?

If possible, it would also help if you could attach a screenshot of an example of this problem.

**Comment by AceDong0803** [comment]
Thanks i am using the newest version of tangram es and ios13 ios14 。the word 华 miss glyph。can try on draw debug text

**Comment by AceDong0803** [comment]
> Hi, please verify some info so that I can help figure this out.
> 
> 
> 
> 1. What version of Tangram ES are you using?
> 
> 2. What version of iOS are you using?
> 
> 3. What scene file are you using?
> 
> 
> 
> If possible, it would also help if you could attach a screenshot of an example of this problem.

I use custom scene file. If necessary i can send it or take screenshot. Thanks a lot. Use xcode 11 simulator Also can show the issue

**Comment by matteblair** [comment]
A screenshot would be helpful! And you don't need to send your whole scene file, but please share what fonts your scene is using.

**Comment by AceDong0803** [comment]
> A screenshot would be helpful! And you don't need to send your whole scene file, but please share what fonts your scene is using.

![image](https://user-images.githubusercontent.com/17538276/93287353-e0ea4680-f80b-11ea-8437-f81c0d3456a8.png)

I just use system font

**Comment by AceDong0803** [comment]
> A screenshot would be helpful! And you don't need to send your whole scene file, but please share what fonts your scene is using.

![image](https://user-images.githubusercontent.com/17538276/93290887-522df780-f814-11ea-9cf8-0236a7260387.png)
![image](https://user-images.githubusercontent.com/17538276/93290901-578b4200-f814-11ea-95c5-0ce957ea94e2.png)

just use the demo app，I add the debug text = "中中" that is Ok when I change text = "华华" that can't render . show missing glyphs. I Use xcode 11.6 simulator. also have the issue in my own phone iOS13.6

**Comment by AceDong0803** [comment]
> A screenshot would be helpful! And you don't need to send your whole scene file, but please share what fonts your scene is using.

**Comment by AceDong0803** [comment]
> A screenshot would be helpful! And you don't need to send your whole scene file, but please share what fonts your scene is using.

![image](https://user-images.githubusercontent.com/17538276/93350933-b62fda80-f86b-11ea-85a6-f487f4b7bd28.png)


I Try to use the debug code then the "华" can render but show strange，the Character "华" should be "PingFangSC-Regular" font in iOS but "PingFangSC-Regular" font in harbuzz code dont call _hb_coretext_aat_shape function. I debug try to call this function directly, the "华" can render。but show very strange

![image](https://user-images.githubusercontent.com/17538276/93351938-e035cc80-f86c-11ea-8f47-91a0eb6a302b.png)

**Comment by matteblair** [comment]
This is very strange! The example you showed with the "华" not rendering is very convincing. I will try to reproduce this locally and debug it.

**Comment by tallytalwar** [comment]
@matteblair could be shot in the dark, but possible place to look at is alfons changes which were done to use coretext on apple platforms: https://github.com/hjanetzek/alfons/pull/14.

**Comment by matteblair** [comment]
I can reproduce this behavior on iOS just as described. The same thing also happens on macOS, like the screenshot below. This confirms that the issue is in an Apple-specific code path, as we suspected. 
<img width="1844" alt="Screen Shot 2020-09-21 at 3 03 19 AM" src="https://user-images.githubusercontent.com/3371850/93740391-5fead080-fbb8-11ea-8f52-881f8ae05d65.png">
From stepping through the code it's not easy to tell exactly where things are going wrong (unfortunately the relevant code is very complex and not very familiar to me!). The first thing I'd like to try is updating the Harfbuzz, ICU, and FreeType libraries like Varun mentioned.

**Comment by AceDong0803** [comment]
> I can reproduce this behavior on iOS just as described. The same thing also happens on macOS, like the screenshot below. This confirms that the issue is in an Apple-specific code path, as we suspected.
> <img alt="Screen Shot 2020-09-21 at 3 03 19 AM" width="1844" src="https://user-images.githubusercontent.com/3371850/93740391-5fead080-fbb8-11ea-8f52-881f8ae05d65.png">
> From stepping through the code it's not easy to tell exactly where things are going wrong (unfortunately the relevant code is very complex and not very familiar to me!). The first thing I'd like to try is updating the Harfbuzz, ICU, and FreeType libraries like Varun mentioned.

yeah,maybe update Harfbuzz, ICU, and FreeType libraries can solve the problem

**Comment by matteblair** [comment]
Success! It took a bit of work to bring Harfbuzz and the other libraries up to their latest version, but now the character that was missing before is rendered as expected.
<img width="1895" alt="Screen Shot 2020-09-28 at 2 46 35 AM" src="https://user-images.githubusercontent.com/3371850/94399846-1eb56c00-0136-11eb-98ce-b720bfbea084.png">

You can test out the change for yourself on this branch: https://github.com/tangrams/tangram-es/tree/harfbuzz-libraries-update Let me know if you spot any other issues there.

The upgrade was more involved than I expected so before I merge this I'll want to do a little cleanup and testing to make sure I didn't create any regressions.

**Comment by tallytalwar** [comment]
Nice Matt. Were you able to follow the instructions on update-libraries.md? Was it just updating? I thought newer harfbuzz, icu, freetype libraries might require updates to alfons too since that's the main consumer of harfbuzz-icu-freetype. I couldn't trace any updates on alfons!!

**Comment by matteblair** [comment]
@tallytalwar Details on the upgrade here: https://github.com/tangrams/harfbuzz-icu-freetype/pull/10

**Comment by AceDong0803** [comment]
> @tallytalwar Details on the upgrade here: [tangrams/harfbuzz-icu-freetype#10](https://github.com/tangrams/harfbuzz-icu-freetype/pull/10)

I change to the update branch and build with xcode 12.0.1，compile failed，the error issue pic 
![image](https://user-images.githubusercontent.com/17538276/95155091-f5e04700-07c5-11eb-8f08-c1134311c5f3.png)

**Comment by matteblair** [comment]
I encountered the same errors when compiling for iOS simulator in Xcode 11. These 3 issues seem to be a result of the Freetype CMake script failing to find the "fcntl.h" Unix header. I was using CMake 3.17.2 when I got these errors. I tried upgrading CMake to the latest version (3.18.3) and that resolved this error.

After resolving that issue I did hit a linking error with the iOS simulator that I'm still figuring out. However this linking error only occurs when targeting the simulator, not an actual iOS device.
